Reference **[Pandas Home Page](https://www.kaggle.com/learn/pandas)**

---

We will now move on to analysing the dataset of players in the Fifa19 computer game.<br>
Each row is a player with a variety of fileds, both quantitaive and qualitative data. 


In [1]:
import pandas as pd
import numpy as np

fifa = pd.read_csv('data/FIFA19.csv')

Create a list of the column names in the  fifa dataframe.<br>
You can use this to help write later snippets of code

In [2]:
fifa.columns

Index(['Unnamed: 0', 'ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag',
       'Overall', 'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingT

In [3]:
fifa.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number  ...     Penalties     Composure       Marking  \
count   18147.000000  ...  18159.000000  18159.000000  18159.000000   
mean       19.546096  ...     48.548598     58.648274     47.281623   
std        15.947765  ...     15.704053     11.436133     19.904397   
min         1.000000  ...      5.000000      3.000000      3.000000   
25%         8.000000  ...     39.000000     51.000000     30.000000   
50%        17.000000  ...     49.000000     60.000000     53.000000   
75%        26.000000  ...     60.000000     67.000000     64.000000   
max        99.000000  ...     92.000000     96.000000     94.000000   

       StandingTackle  SlidingTackle      GKDiving    GKHandling  \
count    18159.000000   18159.000000  18159.000000  18159.000000   
mean        47.697836      45.661435     16.616223     16.391596   
std         21.664004      21.289135     17.695349     16.906900   
min          2.000000       3.000000      1.000000      1.000000   
25%         27.000000      24.000000      8.000000      8.000000   
50%         55.000000      52.000000     11.000000     11.000000   
75%         66.000000      64.000000     14.000000     14.000000   
max         93.000000      91.000000     90.000000     92.000000   

          GKKicking  GKPositioning    GKReflexes  
count  18159.000000   18159.000000  18159.000000  
mean      16.232061      16.388898     16.710887  
std       16.502864      17.034669     17.955119  
min        1.000000       1.000000      1.000000  
25%        8.000000       8.000000      8.000000  
50%       11.000000      11.000000     11.000000  
75%       14.000000      14.000000     14.000000  
max       91.000000      90.000000     94.000000  

[8 rows x 44 columns]

### Notes on mathematical averages
##### Find the mean, median, mode, and range for the following list of 9 values:
##### 13, 18, 13, 14, 13, 16, 14, 21, 13
<br>
The mean is the usual average, so I'll add and then divide:<br>
<br>
<center>(13 + 18 + 13 + 14 + 13 + 16 + 14 + 21 + 13) ÷ 9 = 15</center><br>
<br>
Note that the mean, in this case, isn't a value from the original list. This is a common result. You should not assume that your mean will be one of your original numbers.<br>
<br>
The median is the middle value, so first I'll have to rewrite the list in numerical order:<br>
<br>
<center>13, 13, 13, 13, 14, 14, 16, 18, 21</center><br>
<br>
There are nine numbers in the list, so the middle one will be the (9 + 1) ÷ 2 = 10 ÷ 2 = 5th number:<br>
<br>
<center>13, 13, 13, 13, 14, 14, 16, 18, 21</center><br>
<br>
So the median is 14.<br>
<br>
The mode is the number that is repeated more often than any other, so 13 is the mode.<br>
<br>
The largest value in the list is 21, and the smallest is 13, so the range is 21 – 13 = 8.<br>
<br>
<center>mean: 15</center><br>
<center>median: 14</center><br>
<center>mode: 13</center><br>
<center>range: 8</center><br>

# Exercises

## 1.

What is the mean, median, mode and range of the players' `Strength` in the `Fifa` DataFrame?

In [2]:
Strength_series = fifa["Strength"]
#print(Strength_series)

mean = Strength_series.mean()
median = Strength_series.median()
mode = Strength_series.mode()
strength_range = Strength_series.max()-Strength_series.min()

print("mean =",mean)
print("median =",median)
print("mode = ",mode)
print("range =",strength_range)


mean = 65.31196651798007
median = 67.0
mode =  0    68.0
Name: Strength, dtype: float64
range = 80.0


## 2. 
What nationalities are represented in the dataset? <br>(use `Nationality`. Your answer should not include any duplicates.)

In [36]:
nationalities =fifa['Nationality'].unique()


# Check your answer
nationalities

array(['Argentina', 'Portugal', 'Brazil', 'Spain', 'Belgium', 'Croatia',
       'Uruguay', 'Slovenia', 'Poland', 'Germany', 'France', 'England',
       'Italy', 'Egypt', 'Colombia', 'Denmark', 'Gabon', 'Wales',
       'Senegal', 'Costa Rica', 'Slovakia', 'Netherlands',
       'Bosnia Herzegovina', 'Morocco', 'Serbia', 'Algeria', 'Austria',
       'Greece', 'Chile', 'Sweden', 'Korea Republic', 'Finland', 'Guinea',
       'Montenegro', 'Armenia', 'Switzerland', 'Norway', 'Czech Republic',
       'Scotland', 'Ghana', 'Central African Rep.', 'DR Congo',
       'Ivory Coast', 'Russia', 'Ukraine', 'Iceland', 'Mexico', 'Jamaica',
       'Albania', 'Venezuela', 'Japan', 'Turkey', 'Ecuador', 'Paraguay',
       'Mali', 'Nigeria', 'Cameroon', 'Dominican Republic', 'Israel',
       'Kenya', 'Hungary', 'Republic of Ireland', 'Romania',
       'United States', 'Cape Verde', 'Australia', 'Peru', 'Togo',
       'Syria', 'Zimbabwe', 'Angola', 'Burkina Faso', 'Iran', 'Estonia',
       'Tunisia', 'Equato

## 3.
How often does each Nationality appear in the dataset? Create a Series `players_per_country` mapping nationalities to the count of players of that Nationality.

In [3]:
players_per_country = fifa.Nationality.value_counts()

# Check your answer
players_per_country

England                 1662
Germany                 1198
Spain                   1072
Argentina                937
France                   914
                        ... 
New Caledonia              1
Fiji                       1
São Tomé & Príncipe        1
United Arab Emirates       1
Botswana                   1
Name: Nationality, Length: 164, dtype: int64

## 4.
Create variable `centered_strength_value` containing a version of the `'Strength'` column with the mean strength subtracted.

(Note: this 'centering' transformation is a common preprocessing step before applying various machine learning algorithms.) 
<br>18202   -18.311967
<br>18203     1.688033
<br>18204   -33.311967
<br>18205   -17.311967
<br>18206    -5.311967

In [47]:
print(fifa["Strength"])
centered_strength_value = fifa["Strength"].map(lambda x : x - mean)
# Check your answer
centered_strength_value

0        59.0
1        79.0
2        49.0
3        64.0
4        75.0
         ... 
18202    47.0
18203    67.0
18204    32.0
18205    48.0
18206    60.0
Name: Strength, Length: 18207, dtype: float64


0        -6.311967
1        13.688033
2       -16.311967
3        -1.311967
4         9.688033
           ...    
18202   -18.311967
18203     1.688033
18204   -33.311967
18205   -17.311967
18206    -5.311967
Name: Strength, Length: 18207, dtype: float64

## 5.
You are scouting for players with much higher potential than they are currently showing.
<br> Create a Series `relative_potential_improvement` that stores each players potential ability `Potential` divided by their current ability `Overall`

In [4]:
relative_potential_improvement = fifa.Potential / fifa.Overall
relative_potential_improvement

0        1.000000
1        1.000000
2        1.010870
3        1.021978
4        1.010989
           ...   
18202    1.382979
18203    1.340426
18204    1.425532
18205    1.404255
18206    1.434783
Length: 18207, dtype: float64

From the Series `relative_potential_improvement`
Find the the player with the greatest_relative_potential_improvment.<br>
Hint - intials are D.C.

In [9]:
greatest_relative_potential_improvement = relative_potential_improvement.idxmax()

fifa.iloc[greatest_relative_potential_improvement]


Unnamed: 0                                                 18072
ID                                                        238758
Name                                                 D. Campbell
Age                                                           17
Photo             https://cdn.sofifa.org/players/4/19/238758.png
                                       ...                      
GKHandling                                                  14.0
GKKicking                                                    9.0
GKPositioning                                               14.0
GKReflexes                                                  11.0
Release Clause                                             €165K
Name: 18072, Length: 89, dtype: object

## 6.
You are managing a team in central Spain that are having an injury crisis.<br>
You need a striker that can hold the ball.
<br>Find players that have a heavier build and play striker.
<br>(use unique() to find the list of possible values in columns that are relevant)
<br><br>Create a sublist of players that are playing with clubs in Madrid (have "Madrid" in the club name)as these will be available quickly